slatedb/src/dispatcher.rs

```diff
@@ -0,0 +1,813 @@
+//! # use crate::error::SlateDBError;
+//! # use crate::clock::DefaultSystemClock;
+//! # use crate::watchable_once_cell::WatchableOnceCell;
+//! # use tokio::sync::mpsc;
+//! # use tokio_util::sync::CancellationToken;
+//! # use futures::stream::BoxStream;
+//! # #[tokio::main]
+//! # async fn main() {
+//! enum Message {
+//!     Say(String),
+//! }
+//!
+//! struct PrintMessageHandler;
+//!
+//! #[async_trait::async_trait]
+//! impl MessageHandler<Message> for PrintMessageHandler {
+//!     async fn handle(
+//!         &mut self,
+//!         message: Message,
+//!     ) -> Result<(), SlateDBError> {
+//!         match message {
+//!             Message::Say(msg) => println!("{}", msg),
+//!         }
+//!         Ok(())
+//!     }
+//!
+//!     async fn tickers(&mut self) -> Vec<(Duration, Box<MessageFactory<Message>>)> {
+//!         let mut tickers = Vec::new();
+//!         tickers.push((Duration::from_secs(1), Box::new(|| Message::Say("tick".to_string()))));
+//!         tickers
+//!     }
+//!
+//!     async fn cleanup(
+//!         &mut self,
+//!         messages: BoxStream<'async_trait, T>,
+//!         result: Result<(), SlateDBError>,
+//!     ) -> Result<(), SlateDBError> {
+//!         match result {
+//!             Ok(_) | Err(SlateDBError::BackgroundTaskShutdown) => {
+//!                 messages.for_each(|m| self.handle(m)).await;
+//!             },
+//!             // skipping drain messages on unclean shutdown
+//!             _ => {},
+//!         }
+//!         Ok(())
+//!     }
+//! }
+//!
+//! let clock = DefaultSystemClock::default();
+//! let cancellation_token = CancellationToken::new();
+//! let (tx, rx) = mpsc::unbounded_channel();
+//! let error_state = WatchableOnceCell::new(None);
+//! let dispatcher = MessageDispatcher::new(
+//!     Box::new(PrintMessageHandler),
+//!     rx,
+//!     clock,
+//!     cancellation_token,
+//!     error_state,
+//! );
+//! let join_handle = tokio::spawn(async move { dispatcher.run().await });
+//! tx.send(Message::Say("hello".to_string())).await;
+//! cancellation_token.cancel();
+//! join_handle.await;
+//! # }
+//! ```
+
+// TODO Remove once we've migrated to MessgaeDispatcher
+#![allow(dead_code)]
+
+use std::{future::Future, pin::Pin, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use fail_parallel::{fail_point, FailPointRegistry};
+use futures::{future::select_all, stream::BoxStream};
+use log::warn;
+use tokio::sync::mpsc;
+use tokio_util::sync::CancellationToken;
+
+use crate::{
+    clock::{SystemClock, SystemClockTicker},
+    error::SlateDBError,
+    utils::WatchableOnceCell,
+};
+
+/// A factory for creating messages when a [MessageDispatcherTicker] ticks.
+pub(crate) type MessageFactory<T> = dyn Fn() -> T + Send;
+
+/// A dispatcher that invokes [MessageHandler] callbacks when events occur.
+///
+/// [MessageDispatcher] receives a [MessageHandler] and optional
+/// [mpsc::UnboundedReceiver<T>]. Messages sent to the receiver are passed to the
```

Comment:
is the `rx` optional? I remember seeing that in the original PR but I don't see that one here (and also the docs for `new` need to be updated if that's the case)

Reply:
It was. I got rid of it. All loops except the GC require it, so I went with simplifying the constructor (now that we have fp_registry). I suppose the other option is a builder pattern, which seemed rather overkill. WDYT? I can add an Option<rx>, builder pattern, more constructors, or... :)

Reply:
🫡 I prefer it this way, let's just have the GC loop pass in a no-op one.

Reply:
Fixed docs: 453a639895ff1fa659c2c9ca54034a96cd793796